applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/differentiation_utility.cpp

```diff
@@ -0,0 +1,334 @@
+            Vector RHS_unperturbed;
+            Vector RHS_perturbed;
+            ProcessInfo copy_process_info = rCurrentProcessInfo;
+
+            const SizeType number_of_nodes = rElement.GetGeometry().PointsNumber();
+            const SizeType dimension = rCurrentProcessInfo.GetValue(DOMAIN_SIZE);
+
+            // compute RHS before perturbion
+            rElement.CalculateRightHandSide(RHS_unperturbed, copy_process_info);
+
+            if ( (rOutput.size1() != dimension * number_of_nodes) || (rOutput.size2() != RHS_unperturbed.size() ) )
+                rOutput.resize(dimension * number_of_nodes, RHS_unperturbed.size());
+
+            IndexType index = 0;
+            for(auto& node_i : rElement.GetGeometry())
+            {
+                for(IndexType coord_dir_i = 0; coord_dir_i < dimension; ++coord_dir_i)
+                {
+                    // perturb the design variable
+                    node_i.GetInitialPosition()[coord_dir_i] += rPertubationSize;
+                    node_i[coord_dir_i] += rPertubationSize;
+
+                    // compute RHS after perturbation
+                    rElement.CalculateRightHandSide(RHS_perturbed, copy_process_info);
+
+                    //compute derivative of RHS w.r.t. design variable with finite differences
+                    for(IndexType i = 0; i < RHS_perturbed.size(); ++i)
+                        rOutput( (coord_dir_i + index*dimension), i) = (RHS_perturbed[i]-RHS_unperturbed[i])/rPertubationSize;
+
+                    // Reset perturbed vector
+                    noalias(RHS_perturbed) = ZeroVector(RHS_perturbed.size());
+
+                    // unperturb the design variable
+                    node_i.GetInitialPosition()[coord_dir_i] -= rPertubationSize;
+                    node_i[coord_dir_i] -= rPertubationSize;
+
+                }
+                index++;
+
+                //call one last time to make sure everything is as it was before TODO improve this..
+                rElement.CalculateRightHandSide(RHS_perturbed, copy_process_info);
+
+            }// end loop over element nodes
+        }
+        else
+        {
+            KRATOS_WARNING("DifferentiationUtility") << "Unsupported nodal design variable: " << rDesignVariable << std::endl;
+            if ( (rOutput.size1() != 0) || (rOutput.size2() != 0) )
+                rOutput.resize(0,0,false);
+        }
+
+        KRATOS_CATCH("");
+    }
+
+        void DifferentiationUtility::CalculateRigthHandSideDerivative(Element& rElement,
+                                                const array_1d_component_type& rDesignVariable,
+                                                Node<3>& rNode,
+                                                const double& rPertubationSize,
+                                                Vector& rOutput,
+                                                const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+
+        #ifdef _OPENMP
```

Comment:
ouch
please do not use this
use omp critical or omp single for what you want to achieve => we should probably talk in person how to do this right
or in general use the omp-utils in the core

Reply:
the windows build fails, I think it is because of this
OMP and windows is a sensible issue :)

Reply:
omp critical would work. However we introduced this error to directly hinder someone of calling this from within a parallel loop.

Reply:
but this way you enforce running with one thread even though the rest would work in parallel
If an operation is not threadsafe then using a critical / atomic / single or sth like this would be the proper way

We should talk abt this